### PR TITLE
Avoid unnecessary locks in ROOT function GetBaseClassOffset()

### PR DESCRIPTION
**Pull request by wmtan**
The ROOT function GetBaseClassOffset() does locking that seriously hurts threading efficiency, especially in analysis jobs using pat::. In the vast majority of the cases (in fact in all cases so far observed), the actual offset to the base class object from the derived class object was 0.
This PR replaces  GetBaseClassOffset() with a call to a function of a class template  that returns 0 by default. In cases where the actual offset is non-zero, the class can be specialized to return the proper value. In cases where the true offset is non-zero, and the class has not been properly specialized, a segfault should result, which will be noticed.

### CODE REVIEW

**Comment by wmtan** [comment]
@davidlange6 The changes to analysis code consist entirely of adding two missing std includes that were made visible by removing unnecessary includes in the framework. So, this PR should not be held up by a late analysis signature.

File: DataFormats/Common/test/testOwnVector.cc [hunk 1]
```diff
@@ -2,6 +2,7 @@
 #include <algorithm>
 #include <iterator>
 #include <iostream>
+#include <string.h>
```
Comment (Dr15Jones) [inline DataFormats/Common/test/testOwnVector.cc]: Why the C header? Wouldn't ```<cstring>``` be better?

File: DataFormats/PatCandidates/src/PackedTriggerPrescales.cc [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #include "DataFormats/PatCandidates/interface/PackedTriggerPrescales.h"
 #include "DataFormats/Common/interface/RefProd.h"
+#include <string.h>
```
Comment (Dr15Jones) [inline DataFormats/PatCandidates/src/PackedTriggerPrescales.cc]: Also here is the C header.

File: FWCore/Utilities/interface/OffsetToBase.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/*
+ * For any class used in Views, RefToBase, or Ptr,
+ * class template OffsetToBase must be specialized for any class
+ * with multiple inheritance (i.e. with non-zero offset to
+ * at least one base class).
+ * A specialization would look something like this
+ * (in YourClass.h"
+ *
+ * #include "FWCore/Utilities/interface/OffsetToBase.h
```
Comment (Dr15Jones) [inline FWCore/Utilities/interface/OffsetToBase.h]: Missing a close quote :)

File: FWCore/Utilities/interface/OffsetToBase.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * A specialization would look something like this
+ * (in YourClass.h"
+ *
+ * #include "FWCore/Utilities/interface/OffsetToBase.h 
+ * namespace edm {
+ *   template<>
+ *   class OffsetToBase<YourClass> {
+ *     public OffsetToBase() {}
+ *     size_t offsetToBase(std::type_info const& baseTypeInfo) const {
+ *       YourClass object;
```
Comment (Dr15Jones) [inline FWCore/Utilities/interface/OffsetToBase.h]: Calling a constructor each time this happens isn't a great idea. How about cheating instead:
``` int const dummy = 0; //Just need an adress of something on the stack
    YourClass const* object = reinterpret_cast<YourClass const*>(&dummy);
```

File: DataFormats/Common/test/testOwnVector.cc (hunk 1 above)
Comment (wmtan) [inline DataFormats/Common/test/testOwnVector.cc]: Yes, <cstring> is better.

File: DataFormats/PatCandidates/src/PackedTriggerPrescales.cc (hunk 2 above)
Comment (wmtan) [inline DataFormats/PatCandidates/src/PackedTriggerPrescales.cc]: Yes, cstring is better.

File: FWCore/Utilities/interface/OffsetToBase.h (hunk 3 above)
Comment (wmtan) [inline FWCore/Utilities/interface/OffsetToBase.h]: Will fix.

**Comment by wmtan** [comment]
@cmsbuild please test

**Comment by Dr15Jones** [comment]
+1

**Comment by Dr15Jones** [comment]
@smuzaffar is this test still running?

**Pull request merged**